File: backend/src/bajo_nivel/notaciones/binario/conversion.c
#include <stdio.h>
#include <math.h>

/* Compilación linkeando con el cabecero <math.h>
    gcc conversion.c -o conversion -lm
*/

// =================================================

/**
 * Convierte un número entero en formato decimal
 *    en su equivalente binario
 * @param  num Número decimal entero a convertir
 * @return     Devuelve el número pasado como
 *                 argumento en formato binario
 */
long long decimalToBinary(int num) {
    long long response = 0;
    int resto, pos = 1;
    // paso = 1;   // Decomenta para ver el proceso

    while (num!=0)
    {
        resto = num % 2;
        // Descomenta la siguiente línea para ver el proceso:
        //printf("Paso %d: %d/2, Resto = %d, Cociente = %d\n", paso++, num, resto, num/2);
        num /= 2;
        response += resto * pos;
        pos *= 10;
    }
    return response;
}

/**
 * Convierte un número en formato binario en su equivalente
 *     en decimal como entero.
 * @param  num Número binario a convertir en decimal.
 * @return   Devuelve el número binario pasado como argumento
 *               en su formato decimal.
 */
int binaryToDecimal(long long num) {
    int response = 0, exp = 0, resto;
    while (num!=0)
    {
        resto = num % 10;
        num /= 10;
        response += resto*pow(2,exp);
        ++exp;
    }
    return response;
}

// =============================================================

int main() {
    int numero = 56, binario, decimal;

    // Conversión de decimal a binario
    binario = decimalToBinary(numero);
    printf("El número %d en binario es %d\n", numero, binario);

    // Conversión de binario a decimal
    decimal = binaryToDecimal(binario);
    printf("El número %d en decimal es %d\n", binario, decimal);

    return 0;
}

File: backend/src/bajo_nivel/notaciones/binario/conversion.go
package main

import (
    "fmt"
    "math"
)

/**
 * Convierte un número entero en formato decimal
 *    en su equivalente binario
 * @param  num Número decimal entero a convertir
 * @return     Devuelve el número pasado como
 *                 argumento en formato binario
 */
func decimalToBinary(num int) int {
    response, resto, pos := 0, 0, 1
    // paso = 1;   // Descomenta para ver el proceso

    for {
        resto = num % 2
        // Descomenta la siguiente línea para ver el proceso:
        //printf("Paso %d: %d/2, Resto = %d, Cociente = %d\n", paso++, num, resto, num/2);
        num /= 2
        response += resto * pos
        pos *= 10
        if num == 0 {
            break
        }
    }
    return response;
}

/**
 * Convierte un número en formato binario en su equivalente
 *     en decimal como entero.
 * @param  num Número binario a convertir en decimal.
 * @return   Devuelve el número binario pasado como argumento
 *               en su formato decimal.
 */
func binaryToDecimal(num int) int {
    response, exp, resto := 0, 0, 0
    for {
        resto = num % 10
        num /= 10
        response += resto*int(math.Pow(2.0, float64(exp)))
        exp++
        if num == 0 {
            break
        }
    }
    return response
}

func main() {
    var numero int = 56

    binario := decimalToBinary(numero)
    fmt.Printf("El número %d en binario es %v\n",
               numero, binario)     // 10

    decimal := binaryToDecimal(numero)
    fmt.Printf("El número %d en decimal es %v\n",
               numero, decimal)  // 8
}
